src/main/java/eu/europa/ted/efx/model/expressions/Expression.java

```diff
@@ -0,0 +1,85 @@
+package eu.europa.ted.efx.model.expressions;
+
+import java.lang.reflect.Constructor;
+
+import org.antlr.v4.runtime.misc.ParseCancellationException;
+
+import eu.europa.ted.efx.model.CallStackObject;
+
+public interface Expression extends CallStackObject {
+
+  public String getScript();
+
+  public Boolean isLiteral();
+
+  static <T extends Expression> T instantiate(String script, Class<T> type) {
+    return Expression.instantiate(script, false, type);
+  }
+
+  static <T extends Expression> T from(Expression source, Class<T> returnType) {
+    return Expression.instantiate(source.getScript(), source.isLiteral(), returnType);
+  }
+
+  static <T extends Expression> T instantiate(String script, Boolean isLiteral, Class<T> type) {
+    try {
+      if (isLiteral) {
+        Constructor<T> constructor = type.getConstructor(String.class, Boolean.class);
+        return constructor.newInstance(script, isLiteral);
+      } else {
+        Constructor<T> constructor = type.getConstructor(String.class);
+        return constructor.newInstance(script);
+      }
+    } catch (Exception e) {
+      throw new ParseCancellationException(e);
+    }
+  }
+
+  static <T extends Expression> T empty(Class<T> type) {
+    return instantiate("", type);
+  }
+
+  /**
+   * Base class for all {@link Expression} implementations.
+   */
+  public abstract class Impl implements Expression {
```

Comment:
There are several things I find a bit strange here, and for the other similar `*Expression` interfaces:
1/ having an abstract class named "Impl"
2/ having the implementation of an interface as an inner class of that interface (the usual reasons for using nested classes do not seem fulfilled)
3/ having several classes with the same name "Impl" that are only distinguished by their outer class

My suggestion:
Move the "Impl" to their own file/class, and rename them to "Abstract[InterfaceName]", so "AbstractExpression", "AbstractScalarExpression", etc.

It would also be useful to know why we need interfaces **and** abstract classes.

Reply:
I had two options: Either the classical Java option "interface Expression" and "class ExpressionImpl" separatelly, (which I find rediculous), or find a more creative solution. I did not want use C# convetions (which make much more sense than Java conentions for naming interfaces), so I decided to try something new... We can always revert that to the standard rediculous Java conventions. This is no less rediculous. It just reduces the number of files and the palces to look for some basic stuff: Instead of having to find a java file with a rediculous name for the base implementation of the interface, you have it right there in front of you. Most interfase implementations only have a cosntructor anyway because most of the interfaces are marker interfaces (in lack of a proper type system where we would be able to use actual generics and would not need all these marker interfaces to figure out what is the type of everything).

Reply:
I do not want to rename the interface to AbstractSomething. It is not Abstract. It is the interface that is supposed to be used. This is one more reason that tried to hide the base implementation. I don't want to pass around parameters of type "PathExpressionImpl" or "PathExpression.Impl" or even "AbstractPathExpression". I want to pass around parameters of type PathExpression and I want PathExpression to be an iterface.